<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>UniProt &#8212; CoDIAC 1 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b76e3c8a" />
    <link rel="stylesheet" type="text/css" href="_static/classic.css?v=514cf933" />
    
    <script src="_static/documentation_options.js?v=29a6c3e3"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="PDB" href="PDB.html" />
    <link rel="prev" title="Installation" href="Installation.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="PDB.html" title="PDB"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="Installation.html" title="Installation"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">CoDIAC 1 documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">UniProt</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="module-CoDIAC.UniProt">
<span id="uniprot"></span><h1>UniProt<a class="headerlink" href="#module-CoDIAC.UniProt" title="Link to this heading">¶</a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="CoDIAC.UniProt.makeRefFile">
<span class="sig-prename descclassname"><span class="pre">CoDIAC.UniProt.</span></span><span class="sig-name descname"><span class="pre">makeRefFile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">Uniprot_IDs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outputFile</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/CoDIAC/UniProt.html#makeRefFile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#CoDIAC.UniProt.makeRefFile" title="Link to this definition">¶</a></dt>
<dd><p>Makes a Domain Reference file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>Uniprot_IDs</strong><span class="classifier">list</span></dt><dd><p>list of Uniprot Accession IDs generated by InterPro.py</p>
</dd>
<dt><strong>outputFile</strong><span class="classifier">string</span></dt><dd><p>name of the output file</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>Domain reference file in .csv format</dt><dd></dd>
</dl>
</dd>
<dt class="field-odd">Attributes<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>UniProt_ID</strong><span class="classifier">Uniprot Accession ID</span></dt><dd></dd>
<dt><strong>Gene</strong><span class="classifier">gene name</span></dt><dd></dd>
<dt><strong>Species</strong><span class="classifier">scientific name</span></dt><dd></dd>
<dt><strong>Domains</strong><span class="classifier">Reference domain names with boundary ranges</span></dt><dd></dd>
<dt><strong>Ref Sequence</strong><span class="classifier">Reference sequence</span></dt><dd></dd>
<dt><strong>PDB IDs</strong><span class="classifier">All PDB IDs linked to specific UniProt ID</span></dt><dd></dd>
<dt><strong>Domain Architecture</strong><span class="classifier">Domains found within the protein sequence arranged from N ter to C ter</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CoDIAC.UniProt.print_domain_fasta_file">
<span class="sig-prename descclassname"><span class="pre">CoDIAC.UniProt.</span></span><span class="sig-name descname"><span class="pre">print_domain_fasta_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reference_csv</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Interpro_ID</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_term_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">c_term_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">APPEND</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/CoDIAC/UniProt.html#print_domain_fasta_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#CoDIAC.UniProt.print_domain_fasta_file" title="Link to this definition">¶</a></dt>
<dd><p>Given a Uniprot Reference File of proteins, which contain Interpro domain annotations, create a fasta 
file of the domains of interest found in the reference file. The n_term and c_term offsets will 
build a small padding in case of domain boundary issues (default are 0, but can be set to lengths up to 20)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>reference_csv</strong><span class="classifier">string</span></dt><dd><p>File location that contains the reference of interest (like produced from Uniprot.makeRefFile)</p>
</dd>
<dt><strong>Interpro_ID: string</strong></dt><dd><p>Interpro ID - for example in a reference line such as 
SH3_domain:IPR001452:82:143; SH2:IPR000980:147:246; Prot_kinase_dom:IPR000719:271:524
the interpro ID for the SH3_domain is IPR001452; for the SH2 domain is IPR000980</p>
</dd>
<dt><strong>output_file: string</strong></dt><dd><p>location of output fasta. Fasta headers will be uniprot_ID|InterproID|domain_name|domain_number|domain_start|domain_end
domain number will indicate in proteins with more than one domain of the same type, the occurrence of this domain
from N-to-C numbering. Domain start and stop are relative to ones-based counting of the Uniprot
reference protein</p>
</dd>
<dt><strong>n_term_offset: int</strong></dt><dd><p>Number of amino acids to extend in the n-term direction (up to start of protein)</p>
</dd>
<dt><strong>c_term_offset: int</strong></dt><dd><p>Number of amino acids to extend in the c-term direction (up to end of protein)</p>
</dd>
<dt><strong>APPEND: bool</strong></dt><dd><p>if APPEND=true then it will open an existing file and append, else it overwrites</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>domain_fasta_dict: dict</dt><dd><p>As described in make_domain_fasta_dict - headers are fasta headers and values are the domain sequence
Also writes to output_file</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CoDIAC.UniProt.translate_fasta_back">
<span class="sig-prename descclassname"><span class="pre">CoDIAC.UniProt.</span></span><span class="sig-name descname"><span class="pre">translate_fasta_back</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fasta_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">header_trans_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_fasta</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/CoDIAC/UniProt.html#translate_fasta_back"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#CoDIAC.UniProt.translate_fasta_back" title="Link to this definition">¶</a></dt>
<dd><p>Assuming that you have a fasta_file with shortened headers and would like to move those back 
to the long form names, found in the mapping file (header_trans_file) created by translate_fasta_to_new_headers
Use this function to print at the output_fasta location the fasta file with long headers. You would do this assuming 
you wish to preserve a change, such as through alignment, of the shortened headers.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>fasta_file: str</strong></dt><dd><p>location of input fasta file to translate</p>
</dd>
<dt><strong>header_trans_file: str</strong></dt><dd><p>location that stores the header translation (as written in translate_fasta_to_new_headers)</p>
</dd>
<dt><strong>output_fasta: str</strong></dt><dd><p>location to print the output fasta, using the longer headers</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>No returns, prints by non-append to output_fasta</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CoDIAC.UniProt.translate_fasta_to_new_headers">
<span class="sig-prename descclassname"><span class="pre">CoDIAC.UniProt.</span></span><span class="sig-name descname"><span class="pre">translate_fasta_to_new_headers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fasta_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_fasta</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_array_order</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/CoDIAC/UniProt.html#translate_fasta_to_new_headers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#CoDIAC.UniProt.translate_fasta_to_new_headers" title="Link to this definition">¶</a></dt>
<dd><p>UniProt.print_domain_fasta_file
prints a highly informative fasta header in the production of printing fasta sequences
for a given protein domain of interest. This structure is, in the following order, separated
by ‘|’
uniprot_id, gene_name, domain_name, domainNum, Interpro_ID, start, end</p>
<p>If you wish to change the order, keeping track of this change, use this function to do so
where you will use key_array_order as a list to indicate what items in what listing. 
possible_values = [‘uniprot’, ‘gene’, ‘domain_name’, ‘domain_num’, ‘Interpro_ID’, ‘start’, ‘end’]</p>
<p>For example, if you wanted to use uniprot ID first, gene, and the starting and ending position of the domain you would pass in for 
key_array_order [‘uniprot’, ‘gene’, ‘start’, ‘end’]</p>
<p>This will print the new fasta file at output_fasta and a mapping file, using the base of the output_fasta with _mapping.csv</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>fasta_file: str</strong></dt><dd><p>location of input fasta file to translate</p>
</dd>
<dt><strong>output_fasta: str</strong></dt><dd><p>location to print the output fasta, using the longer headers</p>
</dd>
<dt><strong>key_array_order: list</strong></dt><dd><p>List of strings that includes the order and the values to keep from the possible values
[‘uniprot’, ‘gene’, ‘domain_name’, ‘domain_num’, ‘Interpro_ID’, ‘start’, ‘end’]</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>output_fasta: str</dt><dd><p>Confirmation of location of output file</p>
</dd>
<dt>mapping_file: str</dt><dd><p>Location of the mapping file created. This uses the same base name as the output_fasta and adds _mapping.csv</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <div>
    <h3><a href="index.html">Table of Contents</a></h3>
    <ul>
<li><a class="reference internal" href="#">UniProt</a><ul>
<li><a class="reference internal" href="#CoDIAC.UniProt.makeRefFile"><code class="docutils literal notranslate"><span class="pre">makeRefFile()</span></code></a></li>
<li><a class="reference internal" href="#CoDIAC.UniProt.print_domain_fasta_file"><code class="docutils literal notranslate"><span class="pre">print_domain_fasta_file()</span></code></a></li>
<li><a class="reference internal" href="#CoDIAC.UniProt.translate_fasta_back"><code class="docutils literal notranslate"><span class="pre">translate_fasta_back()</span></code></a></li>
<li><a class="reference internal" href="#CoDIAC.UniProt.translate_fasta_to_new_headers"><code class="docutils literal notranslate"><span class="pre">translate_fasta_to_new_headers()</span></code></a></li>
</ul>
</li>
</ul>

  </div>
  <div>
    <h4>Previous topic</h4>
    <p class="topless"><a href="Installation.html"
                          title="previous chapter">Installation</a></p>
  </div>
  <div>
    <h4>Next topic</h4>
    <p class="topless"><a href="PDB.html"
                          title="next chapter">PDB</a></p>
  </div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/UniProt.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="PDB.html" title="PDB"
             >next</a> |</li>
        <li class="right" >
          <a href="Installation.html" title="Installation"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">CoDIAC 1 documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">UniProt</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
    &#169; Copyright 2023, Naegle Lab.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.2.6.
    </div>
  </body>
</html>